### ISSUE TITLE
Only seeing XXX in UI for MASShortcutView

### ISSUE THREAD

**Issue by beaufour** [shpakovski/MASShortcut#74]
I am using MASShortcut v2.3.1 via CocoaPods. When I insert a `MASShortcutView` into my UI it works fine, but all text is shown as `XXX` in the UI. I suspect that `Localizable.strings` is not part of the pod?

I can fix it by manually copying the strings from the `MASShortcut` source into `Localizable.strings` in my own project, but hopefully that shouldn't be necessary :smile: 

Great library btw!

**Comment by shpakovski** [comment]
Indeed, this is a problem, thank you very much for the report! @zoul Will it help if we rename the `Localizable.strings` and use a custom table name?

**Comment by zoul** [comment]
I currently use the localized MASShortcut in a production app and it works correctly, but I use Carthage, not CocoaPods. So I think the problem is in our CocoaPods manifest, probably our `Localized.strings` file is not installed at all. I’ll take a look at it, just don’t know when yet.

**Comment by shpakovski** [comment]
I think this will not work, because `MASShortcut`s version would just rewrite the local project `Localizable.strings` file, there is no “merge policy” for Pods.

**Comment by zoul** [comment]
Oh, I see – this is why I find Carthage better, it just sticks to the framework that’s already defined. Judging by [this blog post](http://yannickloriot.com/2014/02/cocoapods-and-the-localized-string-files/) and [this Stack Overflow thread](http://stackoverflow.com/questions/12874678/localization-and-cocoapods), it looks like we need to create a resource bundle to stick the strings file into?

**Comment by shpakovski** [comment]
Yeah… This is not very convenient for such a small number of strings :) Any better idea? @beaufour?

**Comment by beaufour** [comment]
I really wish I knew more about the inner workings of CocoaPods, but alas I don't.

**Comment by beaufour** [comment]
Unless I've missed something, it seems like it wasn't too much work :)

**Comment by shpakovski** [comment]
Thanks!

**Comment by zoul** [comment]
This is now available in the 2.3.2 release on CocoaPods. @beaufour, can you please verify that the fix works in the version distributed by CocoaPods?

**Comment by beaufour** [comment]
:+1: Works like a charm.

**Comment by shpakovski** [comment]
Thank you guys 👻

**Comment by zoul** [comment]
:+1:

**Comment by jhoughjr** [comment]
This might be related as I'm having reports of users getting some hangs.
I'm using 2.3.2 as a cocoapod.

the crash is :
10/21/15 11:50:25.617 AM com.apple.xpc.launchd[1]: (com.apple.xpc.launchd.oneshot.0x1000002b.GhostNote[19341]) Service exited due to signal: Killed: 9
10/21/15 11:51:20.967 AM GhostNote[24081]: *** -[NSKeyedUnarchiver initForReadingWithData:]: data is NULL
10/21/15 11:52:15.079 AM GhostNote[24081]: *** -[NSBundle initWithURL:]: nil URL argument
10/21/15 11:52:15.081 AM GhostNote[24081]: (
	0   CoreFoundation                      0x00007fff9e4a2bd2 __exceptionPreprocess + 178
	1   libobjc.A.dylib                     0x00007fff8b8b9dd4 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff9e5093fd +[NSException raise:format:] + 205
	3   Foundation                          0x00007fff9cd84ea2 -[NSBundle initWithURL:] + 87
	4   Foundation                          0x00007fff9cd84e38 +[NSBundle bundleWithURL:] + 45
	5   MASShortcut                         0x0000000109875f64 MASLocalizedString + 132
	6   MASShortcut                         0x0000000109879951 -[MASShortcutView drawRect:] + 989
	7   AppKit                              0x00007fff8c750d5e -[NSView _drawRect:clip:] + 3626
	8   AppKit                              0x00007fff8c7a951d -[NSView _recursiveDisplayAllDirtyWithLockFocus:visRect:] + 1873
	9   AppKit                              0x00007fff8c7a98fa -[NSView _recursiveDisplayAllDirtyWithLockFocus:visRect:] + 2862
	10  AppKit                              0x00007fff8c7a98fa -[NSView _recursiveDisplayAllDirtyWithLockFocus:visRect:] + 2862
	11  AppKit                              0x00007fff8c74e487 -[NSView _recursiveDisplayRectIfNeededIgnoringOpacity:isVisibleRect:rectIsVisibleRectForView:topView:] + 838
	12  AppKit                              0x00007fff8c74dc6c -[NSThemeFrame _recursiveDisplayRectIfNeededIgnoringOpacity:isVisibleRect:rectIsVisibleRectForView:topView:] + 334
	13  AppKit                              0x00007fff8c74c077 -[NSView _displayRectIgnoringOpacity:isVisibleRect:rectIsVisibleRectForView:] + 2449
	14  AppKit                              0x00007fff8c747485 -[NSView displayIfNeeded] + 1950
	15  AppKit                              0x00007fff8c746ccc -[NSWindow displayIfNeeded] + 232
	16  AppKit                              0x00007fff8cdc5206 ___NSWindowGetDisplayCycleObserver_block_invoke6323 + 476
	17  AppKit                              0x00007fff8c746667 __37+[NSDisplayCycle currentDisplayCycle]_block_invoke + 738
	18  QuartzCore                          0x00007fff9e96b91d _ZN2CA11Transaction19run_commit_handlersE18CATransactionPhase + 85
	19  QuartzCore                          0x00007fff9e96add8 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 160
	20  QuartzCore                          0x00007fff9e96aa98 _ZN2CA11Transaction6commitEv + 508
	21  QuartzCore                          0x00007fff9e97628f _ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv + 71
	22  CoreFoundation                      0x00007fff9e437e07 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	23  CoreFoundation                      0x00007fff9e437d77 __CFRunLoopDoObservers + 391
	24  CoreFoundation                      0x00007fff9e416d58 CFRunLoopRunSpecific + 328
	25  HIToolbox                           0x00007fff90f35d55 RunCurrentEventLoopInMode + 235
	26  HIToolbox                           0x00007fff90f35a97 ReceiveNextEventCommon + 184
	27  HIToolbox                           0x00007fff90f359cf _BlockUntilNextEventMatchingListInModeWithFilter + 71
	28  AppKit                              0x00007fff8c5eef3a _DPSNextEvent + 1067
	29  AppKit                              0x00007fff8c5ee369 -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454
	30  AppKit                              0x00007fff8c5e2ecc -[NSApplication run] + 682
	31  GhostNote                           0x00000001097e76dd GhostNote + 108253
	32  libdyld.dylib                       0x00007fff8d6cb5ad start + 1

**Comment by jhoughjr** [comment]
at a breakpoint on line 12 of MASLocalization.m bundleURL is nil.

I'm not sure why I don't get this crash and other users do, but perhaps it has to do with their locale.

I just checked and it appears that the bundle it refers to is in the built product inside the MASShortcut.framework that cocoa pods generated.

**Comment by shpakovski** [comment]
Thanks a lot for the problem report!

To me it looks like this code may help:
```obj-c
NSString *MASLocalizedString(NSString *key, NSString *comment) {
    NSBundle *frameworkBundle = [NSBundle bundleForClass:[MASShortcut class]];
#ifdef COCOAPODS
    NSURL *bundleURL = [frameworkBundle URLForResource:@"MASShortcut" withExtension:@"bundle"];
    frameworkBundle = [NSBundle bundleWithURL:bundleURL];
#else
    return [frameworkBundle localizedStringForKey:key value:@"XXX" table:@"Localizable"];
}
```
Can you check if it works please?

**Comment by jhoughjr** [comment]
I believe that is the code that exists in the current latest version.
I found this issue since some users are getting a blocked UI when showing the view that contains the MASShortcutView.
For some reason though, for me it does not hang, but I would see the exception in the debugger.

The URL was nil, and this code seems to resolve properly and get the bundle.

NSString *MASLocalizedString(NSString *key, NSString *comment) {
    NSBundle *frameworkBundle = nil;
#ifdef COCOAPODS
//    NSURL *bundleURL = [[NSBundle mainBundle] URLForResource:@"MASShortcut" withExtension:@"framework"];
//  
    NSURL * bundleURL = [[NSBundle mainBundle] bundleURL];
    NSURL *frameworkBundleURL = [bundleURL URLByAppendingPathComponent:@"Contents/Frameworks/MASShortcut.framework/Versions/A/Resources/MASShortcut.bundle"];
    
    frameworkBundle = [NSBundle bundleWithURL:frameworkBundleURL];
#else
    frameworkBundle = [NSBundle bundleForClass:[MASShortcut class]];
#endif
    return [frameworkBundle localizedStringForKey:key value:@"XXX" table:@"Localizable"];
}



My pod file has the use_frameworks! directive, perhaps that is the difference.
Though that code appears it wouldn’t return a value if COCOAPODS is defined?

**Comment by beaufour** [comment]
@jhoughjr so you are seeing the `bundleURL` == `nil` in your local build? And you build with `use_frameworks!`? (just making sure I fully understand it)

**Comment by jhoughjr** [comment]
Yes I think that is what is happening. 
Currently in specifying a hard coded url gleaned from inspecting the app bundle and the issue seems fixed.

**Comment by beaufour** [comment]
> Yes I think that is what is happening.
> Currently in specifying a hard coded url gleaned from inspecting the app bundle and the issue seems fixed.

Building as a framework is what Carthage is doing as far as I know, so the trick might just to only do the `COCOAPODS` ugliness when using CocoaPods but not when building as a framework. If you force the code to take the `#else` branch does it work for you?

**Comment by kujenga** [comment]
I just ran a `pod update` on my application and am now running into the same issue that @jhoughjr was seeing. I also am using the `use_frameworks!` flag in my Podfile. The previous version I was using was 2.2.0, which worked just fine for me. For now I've just pinned to that specific version in my Podfile to keep things working, but that isn't the greatest solution.

**Comment by zoul** [comment]
Did we find out whether the `use_frameworks!` flag makes `COCOAPODS` false? That would explain the issue.

**Comment by beaufour** [comment]
> Did we find out whether the use_frameworks! flag makes COCOAPODS false? That would explain the issue.

My guess is that it actually makes it `true`, but that `use_frameworks` doesn't build the bundle so it evaluates to nil. But as said, I'm guessing. I have never used `use_frameworks` myself.

**Comment by kujenga** [comment]
I think that `COCOAPODS` is always set to true when `pod install` is run as part of the configuration performed by CocoaPods, regardless of whether you are using frameworks in your application. In looking at my build settings for the application targets, I see that it has been specifically set in both the "Preprocessor Macros" section for Obj-C and the "Other Swift Flags" section for Swift.

In the Cocoapods documentation [1] on using frameworks it seems to suggest that the original use of `[NSBundle bundleForClass:[MASShortcut class]]` pre https://github.com/shpakovski/MASShortcut/pull/75 should work fine with frameworks, however when I manually modify MASLocalization.m to this effect, I run into the original issue @beaufour was seeing with the `XXX` fallback being displayed everywhere.

[1] http://blog.cocoapods.org/CocoaPods-0.36/#caveats

**Comment by zoul** [comment]
I have tried to fix the problem by not guessing which bundle to use and simply using the one that works. Can you all please test the new code if it really works as intended? Sorry for the trouble.

**Comment by beaufour** [comment]
:+1: Works for me (not using `use_frameworks!`). Nice `dispatch_once` pattern.

**Comment by kujenga** [comment]
Sorry for the relayed response. This code does not crash for me, with `use_frameworks!` turned on, but unfortunately I'm back to the old issue of the XXX default value being shown everywhere.

**Comment by zoul** [comment]
So how about now? The current code tests if the CocaPods bundle contains the localized strings. If not, it switches to the framework bundle instead.

**Comment by jhoughjr** [comment]
Sorry everyone, i was very busy the last couple of weeks wrapping up a contract.
I'll test the latest version and let you know tonight.

**Comment by olivierguerriat** [comment]
Another data point: 7fc5e78 work for me (with `use_frameworks!`). Thanks for the fix!

**Comment by zoul** [comment]
This should be now fixed by d32fc18, sorry for the long wait.

**Comment by madrobby** [comment]
I'm on Cocoapods version of MASShortcut v2.3.3 in a RubyMotion project (I'm on RubyMotion v4.12). I see the "XXX" instead of real strings.

1. Has anybody else experienced this with RubyMotion?
2. "XXX" is an extremely bad placeholder because you can't google for it (try it, it's NSFW)
